### DISCUSSION TITLE
Open nodes recursively

### DISCUSSION THREAD

**Discussion by Vichukano** [nvim-neo-tree/neo-tree.nvim#770] [Q&A] [answered]
How to open node tree recursively?

**Comment by miversen33** [comment]
Tagging @farzadmf

I don't know the answer to this but I would like to ask a few clarifying questions (as I don't believe this feature currently exists).

1) How would you like to see this kind of navigation configured?
    - EG, do we only open all the nodes under the current node and don't go deeper? Do we go 2 levels only? Do we do the entire tree under the current node?
2) What would you expect to happen if a node is automatically opened but there is no permission to open that node?
    - I believe right now an error is thrown to neovim. That works fine if you are intentionally opening a single node, but that behavior isn't really suitable for mass automated opening of an undefined amount of nodes
3) What would you expect to happen if "too many nodes" are open?
    - I believe Neo-tree has a hard limit on something like 10,000 nodes being in memory at once. Regardless of what the limit is, at some point memory will be affected. What would you expect to happen if "too many" are opened?
4) How would you like to be able to "stop" this action?
    - I am thinking from a big picture here. Recursively opening `/` on a multi-terabyte system (for example) would literally never finish. How would you like to be able to "cancel" this action? I don't believe this notion is something that is currently designed into neo-tree.
5) What would you expect to happen with link following?
    - An example, lets say a directory has a link that travels up the tree some where ( eg ` /home/user/Downloads -> /home/user/` (for some reason)). That would directly cause an infinite tree expansion, how would you expect to see that kind of situation handled?

Please note, I am not saying this is impossible. Just that there are quite a few considerations to take into account here (expanding an infinite tree can be quite dangerous).

**Reply by farzadmf** [reply]
All points valid and well-thought of. First thing that comes to my mind (literally me thinking for 1 minute):

All these points can be configurable options with "comments" mentioning the _danger_ to the user, and it's up to them to enable it (and possibly crash their editor when they, =go over the limit etc.)

And I think they could be "gracefully" handled, for example; for example, errors saying:
> max limit is 10,000, and we reached it, so not going any further)

But, I agree, if all the points need to be considered, it would make things very complicated.

But, I'm just wondering, given that there's already `expand_all_nodes`, I guess some of those things are already considered? 🤔

**Reply by miversen33** [reply]
I do not know what goes on in the mind of @cseickel , I assume these points were considered when adding that option.

I haven't contributed much to the codebase here, just spent more time than I'd like inside it so I've been doing my best to provide help/maintenance on things :)

This is absolutely the project owners brainchild lol

**Reply by farzadmf** [reply]
Well, I guess an update on this is that it just dumps a stacktrace (which I personally think is a bad UX) when it tries to expand "too many" nodes. It would be much nicer if it gave a more user friendly error saying "max level reached" or something

But, anyway, as I mentioned in my other comment, I personally can't think of a situation where you want to expand _everything_ (I for one haven't had that use case in a few years I've been programming), but (again personally) I did have cases where I wanted to look at _everything_ inside of a specific directory (which usually happens to have a reasonable amount of items, but nested inside a few directory, so I want to be able to see all at a glance)

**Comment by cseickel** [comment, marked as answer]
Unless I'm confused, I'm pretty sure this already exists. The command is called `expand_all_nodes` and it's mentioned in the README but disabled by default because it is obviously very dangerous in the root of a deeply nested directory.

```lua
["Z"] = "expand_all_nodes"
```

**Reply by miversen33** [reply]
For some reason when I was picking through the code, I didn't think to look for the word "expand" lol. I didn't see this or any of the logic around it. This likely answer's OPs question :)

**Reply by farzadmf** [reply]
@miversen33  my main use case would be to run "expand nodes" scoped to the node I'm currently on. I personally haven't had ever the use case of expanding _everything_, and I personally think if the expansion is scoped, it would be way more useful than the "global" expansion

**Reply by ghostbuster91** [reply]
For anyone like me that came here looking for recursive expand of nodes under the cursor here it is: https://github.com/nvim-neo-tree/neo-tree.nvim/issues/445#issuecomment-1399615937 or https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/368
